radix-engine/src/model/method_authorization.rs

```diff
@@ -2,105 +2,163 @@ use crate::errors::RuntimeError;
 use crate::model::Proof;
 use sbor::*;
-use scrypto::prelude::ProofRule;
+use scrypto::math::Decimal;
+use scrypto::prelude::{NonFungibleAddress, ResourceDefId};
+use scrypto::rust::vec::Vec;
 
-/// Snode which verifies authorization of a method call
 #[derive(Debug, Clone, PartialEq, Eq, Hash, TypeId, Encode, Decode)]
-pub enum MethodAuthorization {
-    Protected(ProofRule),
-    Public,
-    Private,
-    Unsupported,
+pub enum HardProofRuleResource {
+    NonFungible(NonFungibleAddress),
+    Resource(ResourceDefId),
+    SoftResourceNotFound,
 }
 
-impl MethodAuthorization {
-    pub fn check(&self, proofs_vector: &[&[Proof]]) -> Result<(), RuntimeError> {
+impl HardProofRuleResource {
+    pub fn proof_matches(&self, proof: &Proof) -> bool {
         match self {
-            MethodAuthorization::Protected(rule) => Self::check_proof_rule(rule, proofs_vector),
-            MethodAuthorization::Public => Ok(()),
-            MethodAuthorization::Private => Err(RuntimeError::NotAuthorized),
-            MethodAuthorization::Unsupported => Err(RuntimeError::UnsupportedMethod),
+            HardProofRuleResource::NonFungible(non_fungible_address) => {
+                let proof_resource_def_id = proof.resource_def_id();
+                proof_resource_def_id == non_fungible_address.resource_def_id()
+                    && match proof.total_ids() {
+                        Ok(ids) => ids.contains(&non_fungible_address.non_fungible_id()),
+                        Err(_) => false,
+                    }
+            }
+            HardProofRuleResource::Resource(resource_def_id) => {
+                let proof_resource_def_id = proof.resource_def_id();
+                proof_resource_def_id == *resource_def_id
+            }
+            HardProofRuleResource::SoftResourceNotFound => false,
         }
     }
 
-    fn check_proof_rule(
-        proof_rule: &ProofRule,
-        proofs_vector: &[&[Proof]],
-    ) -> Result<(), RuntimeError> {
-        match proof_rule {
-            ProofRule::NonFungible(non_fungible_address) => {
-                for proofs in proofs_vector {
-                    for p in proofs.iter() {
-                        let proof_resource_def_id = p.resource_def_id();
-                        if proof_resource_def_id == non_fungible_address.resource_def_id()
-                            && match p.total_ids() {
-                                Ok(ids) => ids.contains(&non_fungible_address.non_fungible_id()),
-                                Err(_) => false,
-                            }
-                        {
-                            return Ok(());
-                        }
-                    }
-                }
-
-                Err(NotAuthorized)
+    pub fn check_has_amount(&self, amount: Decimal, proofs_vector: &[&[Proof]]) -> bool {
+        for proofs in proofs_vector {
+            if proofs
+                .iter()
+                .any(|p| self.proof_matches(p) && p.total_amount() >= amount)
```

Comment:
We need to add a FIXME here.  The logic is wrong and max sum should be used.

Reply:
good point! though I think this is where the composition logic would be most useful.

Reply:
Yeah, we will need to refactor the composition logic in Proof to make it reusable.